crates/turborepo-auth/src/lib.rs

```diff
@@ -12,19 +12,81 @@ mod ui;
 pub use auth::*;
 pub use error::Error;
 pub use login_server::*;
+use turborepo_api_client::TokenClient;
+use turborepo_vercel_api::token::ResponseTokenMetadata;
 
 /// Token is the result of a successful login. It contains the token string and
-/// a boolean indicating whether the token already existed on the filesystem.
-#[derive(Debug)]
+/// potentially metadata about the token.
+#[derive(Debug, Clone)]
 pub enum Token {
+    /// An existing token on the filesystem
     Existing(String),
+    /// A token that was just created, but not yet written to the filesystem
     New(String),
 }
 impl Token {
-    pub fn into_inner(self) -> String {
+    pub fn new(token: String) -> Self {
+        Self::New(token)
+    }
+    pub fn existing(token: String) -> Self {
+        Self::Existing(token)
+    }
+    /// Checks if the token is valid. We do a few checks:
+    /// 1. Fetch the token metadata.
+    /// 2. From the metadata, check if the token is active.
+    /// 3. If the token is a SAML SSO token, check if it's expired.
+    pub async fn is_valid(&self, client: &impl TokenClient) -> Result<bool, Error> {
+        let metadata = self.fetch_metadata(client).await?;
+        let saml_sso = metadata
+            .scopes
+            .iter()
+            .any(|scope| scope.scope_type == "team" && scope.origin == "saml");
+
+        let active = self.is_token_active(&metadata);
+
+        // Not all scopes have an expiration date, so we need to check if all of them
+        // are expired. If there isn't an expiration date, we assume they are infinite
+        // and therefore cannot be expired.
+        let expired = metadata.scopes.iter().all(|scope| match scope.expires_at {
+            Some(expires_at) => expires_at < metadata.active_at,
+            None => false,
+        });
```

Comment:
I might be missing something, but I think this gets checked by `is_token_active`?

Reply:
I think they were checking slightly different things, but after looking at it again I think I can simplify them and the unit test